sdlib/d/gc/extent.d

```diff
@@ -338,8 +356,69 @@ public:
 			return 0;
 		}
 
-		// Decode freespace, found in the final byte (or two bytes) of the alloc:
-		return readPackedFreeSpace(cast(ushort*) (address + slotSize - 2));
+		// Decode freespace, found in the final usable byte (or two bytes) :
+		return readPackedFreeSpace(getFreeSpacePosition(index));
+	}
+
+	/**
+	 * Finalizer Flag features for slabs.
+	 */
+
+	@property
+	bool supportsFinalization() const {
+		return isFinalizableSizeClass(sizeClass);
+	}
+
+	bool hasFinalizer(uint index) {
+		assert(isSlab(), "hasFinalizer accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+
+		return supportsFinalization && finalizationFlags.valueAtAtomic(index);
+	}
+
+	@property
+	ref shared(Bitmap!128) finalizationFlags() {
+		assert(isSlab(), "finalizationFlags accessed on non slab!");
+		assert(supportsFinalization, "finalizationFlags not supported!");
+
+		return _metadata.slabData.slabFlagData.finalizationFlags;
+	}
+
+	@property
+	ulong* last64Bits() {
+		return cast(ulong*) (address + slotSize - 8);
+	}
+
+	void* getFinalizer(uint index) {
+		assert(isSlab(), "getFinalizer accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+		assert(hasFinalizer(index), "No finalizer is set!");
+
+		return cast(void*) (*last64Bits >> 16 & AddressMask);
```

Comment:
Ok, it is in the free space because you shift, but it's not if you don't shift. (and, incidentally, if you don't shift, it where it always was).